File: FindWeather/Lab-8_FindWeather/RealWeatherService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Lab_8_FindWeather
{
    public class RealWeatherService : IWeatherService
    {
        private static bool useWeatherStack = true;
        private static DateTime lastApiCallTime;

        public Weather GetWeather(Location location)
        {
            // Check if the last API call was less than 30 seconds ago
            if (lastApiCallTime != default && (DateTime.Now - lastApiCallTime).TotalSeconds < 30)
            {
                useWeatherStack = !useWeatherStack; // Toggle the API
            }

            // Update the last API call time
            lastApiCallTime = DateTime.Now;

            // Use WeatherStackAPI or OpenWeatherAPI based on the toggle
            if (useWeatherStack)
            {
                return WeatherStackAPI.Get_Weather(location);
            }
            else
            {
                return OpenWeatherAPI.Get_Weather(location);
            }
        }
    }
}

File: FindWeather/Lab_8_(210042107)/src/WeatherService/RealWeatherService.java
package WeatherService;
import Model.Location;
import Model.Weather;
import WeatherAPI.OpenWeatherAPI;
import WeatherAPI.WeatherStackAPI;
import java.io.IOException;
import java.time.Duration;
import java.time.LocalDateTime;

public class RealWeatherService implements IWeatherService
{
    private static boolean useWeatherStack = true;
    private static LocalDateTime lastApiCallTime;

    @Override
    public Weather getWeather(Location location) throws IOException
    {
        if (lastApiCallTime != null && Duration.between(lastApiCallTime, LocalDateTime.now()).getSeconds() < 30)
        {
            useWeatherStack = !useWeatherStack;
        }

        lastApiCallTime = LocalDateTime.now();

        if (useWeatherStack)
        {
            return WeatherStackAPI.Get_Weather(location);
        }

        else
        {
            return OpenWeatherAPI.Get_Weather(location);
        }
    }
}
